### PR TITLE
Add functions to manipulate colshapes parameters

### PR DESCRIPTION
**Pull request by StrixG**
- [x] **Radius** *(circle, sphere, tube)*
```lua
float getColShapeRadius(colshape colShape)
bool setColShapeRadius(colshape colShape, float radius)
```
- [x] **Size**

**Tube (height)**
```lua
float getColShapeSize(colshape colShape)
bool setColShapeSize(colshape colShape, float size)
```
**Rectangle (width, height)**
```lua
float, float getColShapeSize(colshape colShape)
bool setColShapeSize(colshape colShape, Vector2 size)
bool setColShapeSize(colshape colShape, float width, float height)
```
**Cuboid (width, depth, height)**
```lua
float, float, flaot getColShapeSize(colshape colShape)
bool setColShapeSize(colshape colShape, Vector3 size)
bool setColShapeSize(colshape colShape, float width, float depth, float height)
```
- [x] **Polygon points**
```lua
table getColPolygonPoints(colshape colPolygon)
Vector2 getColPolygonPointPosition(colshape colPolygon, int pointIndex)
bool setColPolygonPointPosition(colshape colPolygon, int pointIndex, Vector2 position)
bool addColPolygonPoint(colshape colPolygon, Vector2 position[, int pointIndex])
bool removeColPolygonPoint(colshape colPolygon, int pointIndex)
```

### CODE REVIEW

**Comment by CrosRoad95** [comment]
you should also add heigth for polygons, https://github.com/multitheftauto/mtasa-blue/pull/226

**Comment by StrixG** [comment]
Should I leave one `setColShapeSize` function or make three different functions to change parameters separately (`setColShapeWidth, setColShapeHeight, setColShapeDepth`)?

**Comment by PlatinMTA** [comment]
> Should I leave one `setColShapeSize` function or make three different functions to change parameters separately (`setColShapeWidth, setColShapeHeight, setColShapeDepth`)?

I think setColShapeSize is alright, having to many functions just to change the size of the colshape would be confusing/annoying tbh. If the function adapts to the type of colshape it's changing that's all we need i think

**Comment by StrixG** [comment]
I'm not sure if I should use the sync time context here or not.

**Comment by qaisjp** [comment]
The code looks good to me.

> I'm not sure if I should use the sync time context here or not.

I'm not sure either. Let's figure it out together.

[I found this (introduced in initial commit):](https://github.com/multitheftauto/mtasa-blue/blob/e11685cab4beb7958ab202261f9c9d9b4ce71e58/Client/mods/deathmatch/logic/CClientEntity.h#L166-L171)

```cpp
// This is used for realtime synced elements. Whenever a position/rotation change is
// forced from the server either in form of spawn or setElementPosition/rotation a new
// value is assigned to this from the server. This value also comes with the sync packets.
// If this value doesn't match the value from the sync packet, the packet should be
// ignored. Note that if this value is 0, all sync packets should be accepted. This is
// so we don't need this byte when the element is created first.
```

Markers don't use a sync time context, so I'd guess that colshapes wouldn't need them either. What do you think?

File: Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 1]
```diff
@@ -9481,9 +9477,10 @@ bool CStaticFunctionDefinitions::RemoveColPolygonPoint(CColPolygon* pColPolygon,
     {
         pColPolygon->RemovePoint(uiPointIndex);
 
+        RefreshColShapeColliders(pColPolygon);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:9474]: Nice catch

**Comment by qaisjp** [comment]
have you tested the latest version?

File: Client/mods/deathmatch/logic/CClientColPolygon.cpp [hunk 2]
```diff
@@ -69,20 +69,68 @@ void CClientColPolygon::SetPosition(const CVector& vecPosition)
...
-    float fDistanceX = vecPoint.fX - m_vecPosition.fX;
-    float fDistanceY = vecPoint.fY - m_vecPosition.fY;
+    if (iPointIndex < 0)
+    {
+        m_Points.push_back(vecPoint);
+    }
+    else
+    {
+        if (iPointIndex >= m_Points.size())
+            return;
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CClientColPolygon.cpp]: what do you think about returning false here (and therefore also the lua function)?

File: Client/mods/deathmatch/logic/CClientColPolygon.cpp [hunk 3]
```diff
@@ -69,20 +69,68 @@ void CClientColPolygon::SetPosition(const CVector& vecPosition)
...
-        m_fRadius = fDist;
+        m_fRadius = fDistance;
         SizeChanged();
     }
+}
+
+void CClientColPolygon::RemovePoint(unsigned int uiPointIndex)
+{
+    if (m_Points.size() <= 3)
+        return;
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CClientColPolygon.cpp]: same here about returning false (and for lua) [and elsewhere, for other functions too]

**Review by qaisjp — APPROVED** [review]
I'm overall happy with this and can merge once comments have been addressed.

I don't have time to test, but I'm confident in your work, so I'll merge once you've confirmed you've tested the latest version

**Comment by StrixG** [comment]
These functions also allow you to edit colshapes of markers and pickups. Should I leave this opportunity?

Everything tested. Works good.

**Comment by qaisjp** [comment]
> These functions also allow you to edit colshapes of markers and pickups.

That's funky. Let's leave it in!

**Comment by Dutchman101** [comment]
> Warning	C4018	'>=': signed/unsigned mismatch   \mtasa-blue\Client\mods\deathmatch\logic\CClientColPolygon.cpp:80
> Warning	C4018	'>=': signed/unsigned mismatch\   mtasa-blue\Server\mods\deathmatch\logic\CColPolygon.cpp:96

@StrixG the point index should be unsigned int.. why would you use int? This adds new warnings to our list

**Comment by StrixG** [comment]
I use -1 as default value to add a point to the end of the list. I should have used static_cast on line 80.

**Pull request merged**